File: packages/typescript/src/rag/chunker.ts
/**
 * Document chunking — splits text into overlapping chunks for embedding.
 * Supports: plain text, markdown, code.
 */

export interface ChunkOptions {
  /** Max characters per chunk */
  chunkSize?: number;
  /** Overlap between chunks */
  chunkOverlap?: number;
  /** Split strategy */
  strategy?: "paragraph" | "sentence" | "fixed" | "markdown";
}

export interface TextChunk {
  content: string;
  index: number;
  metadata?: Record<string, unknown>;
}

const PARAGRAPH_SEPARATORS = ["\n\n", "\r\n\r\n"];
const SENTENCE_ENDINGS = /(?<=[.!?])\s+/;
const MARKDOWN_HEADERS = /^#{1,6}\s/m;

export function chunkText(text: string, options: ChunkOptions = {}): TextChunk[] {
  const { chunkSize = 1000, chunkOverlap = 200, strategy = "paragraph" } = options;
  const chunks: TextChunk[] = [];

  if (!text || text.trim().length === 0) return [];

  let segments: string[];

  switch (strategy) {
    case "markdown":
      segments = splitByMarkdown(text);
      break;
    case "sentence":
      segments = text.split(SENTENCE_ENDINGS).filter(s => s.trim().length > 0);
      break;
    case "paragraph":
      segments = splitByParagraph(text);
      break;
    case "fixed":
    default:
      segments = splitFixed(text, chunkSize);
      break;
  }

  // Merge small segments and split large ones
  let currentChunk = "";
  let chunkIndex = 0;

  for (const segment of segments) {
    if (currentChunk.length + segment.length > chunkSize && currentChunk.length > 0) {
      chunks.push({ content: currentChunk.trim(), index: chunkIndex++ });

      // Keep overlap
      if (chunkOverlap > 0) {
        const overlapStart = Math.max(0, currentChunk.length - chunkOverlap);
        currentChunk = currentChunk.slice(overlapStart) + " " + segment;
      } else {
        currentChunk = segment;
      }
    } else {
      currentChunk += (currentChunk ? " " : "") + segment;
    }

    // If single segment exceeds chunk size, force split
    if (currentChunk.length > chunkSize * 1.5) {
      const forceSplit = splitFixed(currentChunk, chunkSize);
      for (let i = 0; i < forceSplit.length - 1; i++) {
        chunks.push({ content: forceSplit[i].trim(), index: chunkIndex++ });
      }
      currentChunk = forceSplit[forceSplit.length - 1];
    }
  }

  if (currentChunk.trim().length > 0) {
    chunks.push({ content: currentChunk.trim(), index: chunkIndex });
  }

  return chunks;
}

function splitByParagraph(text: string): string[] {
  for (const sep of PARAGRAPH_SEPARATORS) {
    if (text.includes(sep)) {
      return text.split(sep).filter(s => s.trim().length > 0);
    }
  }
  return text.split("\n").filter(s => s.trim().length > 0);
}

function splitByMarkdown(text: string): string[] {
  const sections: string[] = [];
  const lines = text.split("\n");
  let current = "";

  for (const line of lines) {
    if (MARKDOWN_HEADERS.test(line) && current.trim().length > 0) {
      sections.push(current.trim());
      current = line;
    } else {
      current += "\n" + line;
    }
  }
  if (current.trim().length > 0) sections.push(current.trim());

  return sections.length > 0 ? sections : splitByParagraph(text);
}

function splitFixed(text: string, size: number): string[] {
  const chunks: string[] = [];
  for (let i = 0; i < text.length; i += size) {
    chunks.push(text.slice(i, i + size));
  }
  return chunks;
}

File: packages/python/bulwark_ai/rag/chunker.py
"""
Document chunking -- splits text into overlapping chunks for embedding.

Supports strategies: paragraph, sentence, markdown, fixed.
"""

from __future__ import annotations

import re
from typing import Literal

from .types import TextChunk


PARAGRAPH_SEPARATORS = ["\n\n", "\r\n\r\n"]
SENTENCE_ENDINGS = re.compile(r"(?<=[.!?])\s+")
MARKDOWN_HEADERS = re.compile(r"^#{1,6}\s", re.MULTILINE)


def _split_by_paragraph(text: str) -> list[str]:
    for sep in PARAGRAPH_SEPARATORS:
        if sep in text:
            return [s for s in text.split(sep) if s.strip()]
    return [s for s in text.split("\n") if s.strip()]


def _split_by_markdown(text: str) -> list[str]:
    sections: list[str] = []
    lines = text.split("\n")
    current = ""

    for line in lines:
        if MARKDOWN_HEADERS.match(line) and current.strip():
            sections.append(current.strip())
            current = line
        else:
            current += "\n" + line

    if current.strip():
        sections.append(current.strip())

    return sections if sections else _split_by_paragraph(text)


def _split_fixed(text: str, size: int) -> list[str]:
    return [text[i : i + size] for i in range(0, len(text), size)]


def chunk_text(
    text: str,
    *,
    chunk_size: int = 1000,
    chunk_overlap: int = 200,
    strategy: Literal["paragraph", "sentence", "fixed", "markdown"] = "paragraph",
) -> list[TextChunk]:
    """
    Split text into overlapping chunks suitable for embedding.

    Args:
        text: The input text to chunk.
        chunk_size: Maximum characters per chunk (default 1000).
        chunk_overlap: Overlap between consecutive chunks (default 200).
        strategy: Split strategy -- paragraph, sentence, markdown, or fixed.

    Returns:
        List of TextChunk objects with content and index.
    """
    if not text or not text.strip():
        return []

    chunks: list[TextChunk] = []

    if strategy == "markdown":
        segments = _split_by_markdown(text)
    elif strategy == "sentence":
        segments = [s for s in SENTENCE_ENDINGS.split(text) if s.strip()]
    elif strategy == "paragraph":
        segments = _split_by_paragraph(text)
    elif strategy == "fixed":
        segments = _split_fixed(text, chunk_size)
    else:
        segments = _split_fixed(text, chunk_size)

    # Merge small segments and split large ones
    current_chunk = ""
    chunk_index = 0

    for segment in segments:
        if len(current_chunk) + len(segment) > chunk_size and current_chunk:
            chunks.append(TextChunk(content=current_chunk.strip(), index=chunk_index))
            chunk_index += 1

            # Keep overlap
            if chunk_overlap > 0:
                overlap_start = max(0, len(current_chunk) - chunk_overlap)
                current_chunk = current_chunk[overlap_start:] + " " + segment
            else:
                current_chunk = segment
        else:
            current_chunk += (" " if current_chunk else "") + segment

        # If single segment exceeds chunk size, force split
        if len(current_chunk) > chunk_size * 1.5:
            force_split = _split_fixed(current_chunk, chunk_size)
            for part in force_split[:-1]:
                chunks.append(TextChunk(content=part.strip(), index=chunk_index))
                chunk_index += 1
            current_chunk = force_split[-1]

    if current_chunk.strip():
        chunks.append(TextChunk(content=current_chunk.strip(), index=chunk_index))

    return chunks
